### PR TITLE
Fix memory leaks from `pylsqpack.Decoder` and `pylsqpack.Encoder`

### PR DESCRIPTION
**Pull request by msoxzw**
Partially fix https://github.com/aiortc/pylsqpack/issues/14

### CODE REVIEW

**Comment by jlaine** [comment]
OK, I now realise I never knew how to properly write a Python object destructor in C. Good catch!

**Comment by msoxzw** [comment]
Since I do not have much time and experiences about Python C extension, I attempt to fix memory leaks that I can deal with, and finding and resolving all memory leaks are far beyond my capability.

Current situation has been mentioned in https://github.com/aiortc/pylsqpack/issues/14#issuecomment-1081023413, and some additional memory leaks are evidenced by [this log](https://github.com/aiortc/aioquic/issues/272) generated by `tracemalloc`.

Furthermore, my workaround probably fix memory leaks in specific conditions. I have observed that memory will be correctly released if stream id remains unchanged, but otherwise not, e.g. `pylsqpack.Decoder.feed_encoder`.

File: src/pylsqpack/binding.c [hunk 1]
```diff
@@ -444,11 +450,13 @@ Encoder_encode(EncoderObject *self, PyObject *args, PyObject *kwargs)
...
-        2,
-        PyBytes_FromStringAndSize((const char*)self->enc_buf, enc_off),
-        PyBytes_FromStringAndSize((const char*)self->hdr_buf + pfx_off, hdr_off - pfx_off)
-    );
+    name =  PyBytes_FromStringAndSize((const char*)self->enc_buf, enc_off);
+    value = PyBytes_FromStringAndSize((const char*)self->hdr_buf + pfx_off, hdr_off - pfx_off);
+    tuple = PyTuple_Pack(2, name, value);
+    Py_DECREF(name);
+    Py_DECREF(value);
+
```
Comment (jlaine) [inline src/pylsqpack/binding.c:458]: We seem to be missing `Py_DECREF(tuple)` too if I understand correctly ?

**Comment by jlaine** [comment]
I *really* appreciate what you're doing here, you've uncovered lots of good stuff already! To be honest my experience with C Python extensions is fairly limited too, so your work is very welcome here.

File: src/pylsqpack/binding.c [hunk 2]
```diff
@@ -271,11 +271,14 @@ Decoder_resume_header(DecoderObject *self, PyObject *args, PyObject *kwargs)
     STAILQ_REMOVE(&self->pending_blocks, hblock, header_block, entries);
     header_block_free(hblock);
 
-    return PyTuple_Pack(
+    tuple = PyTuple_Pack(
         2,
         PyBytes_FromStringAndSize((const char*)self->dec_buf, dec_len),
```
Comment (jlaine) [inline src/pylsqpack/binding.c:276]: I guess we also need to Py_DECREF the bytes objects.

**Comment by msoxzw** [comment]
Many thanks!
No memory leaks occurred after my simple "DNS over HTTP/3" proxy ran for 10 hours or so.

Moreover, I figure out why memory leaks happen only in specific conditions. Python optimization for small integer and small bytes contributes to this phenomenon. The following code snippets illustrate these optimization:

```
a = 256
b = 256
a is b
```
Output: `True`
```
a = 257
b = 257
a is b
```
Output: `False`
```
a = b'x'
b = b'x'
a is b
```
Output: `True`
```
a = b'xx'
b = b'xx'
a is b
```
Output: `False`

Initially, I refrained from adding Py_DECREF in order to avoid severer bugs, unless it indeed fixed certain memory leaks, so Py_DECREF for small integer and small bytes was ignored accordingly. Therefore, it reminds me not to use small integer and small bytes as test data and test results.

**Comment by jlaine** [comment]
Right I did read about this "small int optimization" in the docs. Luckily we just need to Py_DECREF regardless of what Python does under the hood :)

Again a big thank you for your investigations and your fixes! To make sure everyone benefits from your work immediately, the releases currently on PyPI contain your fixes both for `pylsqpack` and `aioquic`.

**Pull request merged**